Documentation/contributing/testing/bpf.rst

```diff
@@ -0,0 +1,307 @@
+        void *data = (void *)(long)ctx->data;
+        void *data_end = (void *)(long)ctx->data_end;
+
+        if (data + sizeof(__u32) > data_end)
+            test_fatal("status code out of bounds");
+
+        __u32 *status_code = data;
+
+        if (*status_code != XDP_TX)
+            test_fatal("status code != XDP_TX");
+
+        data += sizeof(__u32);
+
+        if (data + sizeof(struct ethhdr) > data_end)
+            test_fatal("ctx doesn't fit ethhdr");
+
+        struct ethhdr *l2 = data;
+
+        data += sizeof(struct ethhdr);
+
+        if (memcmp(l2->h_source, fib_smac, sizeof(fib_smac)) != 0)
+            test_fatal("l2->h_source != fib_smac");
+
+        if (memcmp(l2->h_dest, fib_dmac, sizeof(fib_dmac)) != 0)
+            test_fatal("l2->h_dest != fib_dmac");
+
+        if (data + sizeof(struct iphdr) > data_end)
+            test_fatal("ctx doesn't fit iphdr");
+
+        test_finish();
+    }
+
+Function reference
+------------------
+
+* ``test_log(fmt, args...)`` - will write a log message. The conversion specifiers supported by *fmt* are the same as for
+  ``bpf_trace_printk()``. They are **%d**, **%i**, **%u**, **%x**, **%ld**, **%li**, **%lu**, **%lx**, **%lld**, **%lli**, 
+  **%llu**, **%llx**. No modifier (size of field, padding with zeroes, etc.) is available.
+
+* ``test_fail()`` - marks the current test or sub-test as failed but will continue execution.
+
+* ``test_fail_now()`` - marks the current test or sub-test as failed and will stop execution of the 
+  test or sub-test (If called in a sub-tests, the other sub-tests will still run).
+
+* ``test_fatal(fmt, args...)`` - writes a log and then calls ``test_fail_now()``
+
+* ``assert(stmt)`` - asserts that the statement within is true and will ``test_fail_now()`` otherwise.
+  Assert will log the file and line number of the assert statement.
+
+* ``test_skip()`` - marks the current test or sub-test as skipped but will continue execution.
+
+* ``test_skip_now()`` - marks the current test or sub-test as skipped and will stop execution of the 
+  test or sub-test (If called in a sub-tests, the other sub-tests will still run).
+
+* ``test_init()`` - this function initializes internal state and must be called before any of the 
+  functions above can be called.
+
+* ``test_finish()`` - this function submits the results and returns from the current function.
+
+Function mocking
+----------------
+
+Being able to mock out a function is a great tool to have when creating tests for a number of 
+reasons. You might for example want to tests what happens if a specific function returns an error 
+to see if it is handled gracefully. You might want to proxy function calls to record if the function
+under test actually called specific dependencies. Or you might want to test code that uses helpers
+which rely on state we can't set in BPF like the routing table.
+
+Mocking is a fairy easy:
+
+1. Create a macro with the exact same name and make it equal to another function name.
```

Comment:
Not clear. Same name as what? How do you make a macro equal to a function name?

Maybe swap with step 2., and something like “Create a ``#define`` macro named like the original function to replace, but pointing to the newly defined function from step 1.”?